### PR TITLE
fix data processing + resume training

### PR DESCRIPTION
**Pull request by Dogacel**
## Motivation

- Fix training pipeline to resume from the last checkpoint properly.
- Loss tracking on wandb causes gradient accumulation and thus memory leak as wandb commits periodically.
- Track draft configuration on WandB.
- Remove bad data during data preparation to fix zero loss dat points.
- Fix eval data not being cached during first training step.

## Modifications

- Allow loading trainer optimizer states, global step and epoch from the checkpoint.
- Use detached losses for tracking metrics.
- Skip rows that start with assistant rows.
- 

## Accuracy Test

- I have additional changes on my private branch to make it work on our HPC cluster and I am using those changes to train our models. However I would be glad if someone can double-check my patch also works on your setup.
- I've run processing and I see ShareGPT dataset has equal number of train/test data poitns with post-processing (regenerate train data) and with no post-processing.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

**Comment by FrankLeeeee** [comment]
Hi @Dogacel , can you apply code formatting via pre-commit?

**Comment by Dogacel** [comment]
> Hi @Dogacel , can you apply code formatting via pre-commit?

Got it!

The link in PR description is dead, but I assume it is this one -> https://pre-commit.com/#install

I've run `pre-commit run --all-file` and committed.

**Comment by Dogacel** [comment]
Gentle ping @FrankLeeeee

**Comment by FrankLeeeee** [comment]
Hi @Dogacel  sorry for my late reply, your PR looks good to me, can you help to resolve the conflict with the main branch?

**Comment by Dogacel** [comment]
> Hi @Dogacel sorry for my late reply, your PR looks good to me, can you help to resolve the conflict with the main branch?

Done!

I also fixed the bug where eval dataset is not being cached like the train dataset.

For the future, believe we can somehow wrap all the logic inside train_eagle3.py and train_dflash.py to a common wrapper, what do you think?

**Comment by FrankLeeeee** [comment]
This CI failed, can you check the errors?
